components/expression/expression-loader.js

```diff
@@ -52,117 +54,237 @@ genotet.ExpressionLoader.prototype.loadExpressionMatrix_ = function(matrixName,
   };
 
   $.get(genotet.data.serverURL, params, function(data) {
-    // Store the last applied data selectors.
-    _(data).extend({
-      matrixname: matrixName,
-      geneRegex: geneRegex,
-      conditionRegex: conditionRegex
-    });
+      // Store the last applied data selectors.
+      _(data).extend({
+        matrixname: matrixName,
+        geneRegex: geneRegex,
+        conditionRegex: conditionRegex
+      });
 
-    this.data.matrix = data;
+      this.data.matrix = data;
 
-    this.signal('loadComplete');
-  }.bind(this), 'jsonp')
+      this.signal('loadComplete');
+    }.bind(this), 'jsonp')
     .fail(this.fail.bind(this, 'cannot load expression matrix', params));
 };
 
-
-/*
-LoaderHeatmap.prototype.updateData = function(identifier) {
-  if (identifier.action == 'node') {
-    this.loadLine(this.lastIdentifier.mat, identifier.name);
-    if (identifier.net != null) {
-      this.loadHeatmapTargets(identifier.net, identifier.name);
-    }
-  }else if (identifier.action == 'link') {
-    this.clearLines();
-    this.loadLine(this.lastIdentifier.mat, identifier.source);
-    this.loadLine(this.lastIdentifier.mat, identifier.target);
+/**
+ * Updates the genes in the current expression.
+ * @param {string} method Update method, either 'set' or 'add'.
+ * @param {string} geneRegex Regex that selects the genes to be updated.
+ */
+genotet.ExpressionLoader.prototype.updateGenes = function(method, geneRegex, originParams) {
+  var regex = this.data.matrix.geneRegex;
+  geneRegex = geneRegex.toUpperCase();
+  switch (method) {
+    case 'setGene':
+      // Totally replace the regex.
+      regex = geneRegex;
+      this.data.genes = [];
+      this.data.genes.push(geneRegex);
+      break;
+    case 'addGene':
+      // Concat the two regex's. We need to include the previously existing
+      // genes too so as to find the edges between the new genes and old
+      // ones.
+      regex += '|' + geneRegex;
+      this.data.genes.push(geneRegex);
+      break;
+    case 'removeGene':
+      // Remove the regex.
+      var index = this.data.genes.indexOf(geneRegex);
+      if (index == -1) {
+        genotet.error('invalid gene regex', geneRegex);
+        return;
+      };
+      this.data.genes.splice(index, 1);
+      regex = '';
+      this.data.genes.forEach(function(geneRegexs, i) {
+        regex += geneRegexs + (i == this.data.genes.length - 1 ? '' : '|');
+      }, this);
+      break;
   }
+  var conditionRegex = '';
+  this.data.conditions.forEach(function(conditionNames, i) {
+    conditionRegex += conditionNames + (i == this.data.conditions.length - 1 ? '' : '|');
+  }, this);
+  this.signal('loadStart');
+  var params = {
+    type: 'expression',
+    mat: originParams.matrixName,
+    exprows: regex,
+    expcols: conditionRegex
+  };
+  $.get(genotet.data.serverURL, params, function(data) {
```

Comment:
This seems to be exactly the same ajax routine as "load()". If you are creating regex's that represent the new row/column subsets, please re-use load().